src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,281 @@
+#include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
+
+#include "PyDeserializer.hpp"
+
+#include <system_error>
+#include <type_traits>
+#include <utility>
+
+#include <clp/ErrorCode.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ffi/ir_stream/Deserializer.hpp>
+#include <clp/ffi/ir_stream/IrUnitType.hpp>
+#include <clp/ffi/KeyValuePairLogEvent.hpp>
+#include <clp/ffi/SchemaTree.hpp>
+#include <clp/time_types.hpp>
+#include <clp/TraceableException.hpp>
+
+#include <clp_ffi_py/api_decoration.hpp>
+#include <clp_ffi_py/error_messages.hpp>
+#include <clp_ffi_py/ExceptionFFI.hpp>
+#include <clp_ffi_py/ir/native/DeserializerBufferReader.hpp>
+#include <clp_ffi_py/ir/native/error_messages.hpp>
+#include <clp_ffi_py/ir/native/PyKeyValuePairLogEvent.hpp>
+#include <clp_ffi_py/PyObjectCast.hpp>
+#include <clp_ffi_py/utils.hpp>
+
+namespace clp_ffi_py::ir::native {
+using clp::ffi::ir_stream::IRErrorCode;
+using clp::ffi::ir_stream::IrUnitType;
+
+namespace {
+/**
+ * Callback of `PyDeserializer`'s `__init__` method:
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+PyDoc_STRVAR(
+        cPyDeserializerDoc,
+        "Deserializer for deserializing CLP key-value pair IR streams.\n"
+        "This class deserializes log events from a CLP key-value pair IR stream.\n\n"
+        "__init__(self, input_stream, buffer_capacity=65536, allow_incomplete_stream=False)\n\n"
+        "Initializes a :class:`Deserializer` instance with the given inputs. Note that each"
+        " object should only be initialized once. Double initialization will result in a memory"
+        " leak.\n\n"
+        ":param input_stream: Input stream that contains serialized CLP IR.\n"
+        ":type input_stream: IO[bytes]\n"
+        ":param buffer_capacity: The capacity used to initialize the underlying read buffer.\n"
+        ":type buffer_capacity: int\n"
+        ":param allow_incomplete_stream: If set to `True`, an incomplete CLP IR stream is not"
+        " treated as an error.\n"
+        ":type allow_incomplete_stream: bool\n"
+);
+CLP_FFI_PY_METHOD auto
+PyDeserializer_init(PyDeserializer* self, PyObject* args, PyObject* keywords) -> int;
+
+/**
+ * Callback of `PyDeserializer`'s `deserialize_log_event`.
+ */
+PyDoc_STRVAR(
+        cPyDeserializerDeserializeLogEventDoc,
+        "deserialize_log_event(self)\n"
+        "--\n\n"
+        "Deserializes the IR stream until the next log event has been deserialized.\n\n"
```

Comment:
Tbh, the sentence looks confusing to me.
How about?
```suggestion
        "Deserializes the next log event from the IR stream.\n\n"
```

Reply:
It is written this way since the stream may also contains other IR units. This method deserializes all IR units until the next log event IR unit has been deserialized

Reply:
Do customers need to be aware of the IR units? If so then I agree with you.

Reply:
At this point no. Sure, let's take ur suggestion for this release